bus-mapping/src/evm/opcodes/begin_end_tx.rs

```diff
@@ -628,46 +627,107 @@ fn gen_tx_eip2930_ops(
+    state: &mut CircuitInputStateRef,
+    exec_step: &mut ExecStep,
+) -> Result<(), Error> {
+    let tx_id = state.tx_ctx.id();
 
-    // Add copy event for access-list addresses.
+    // Build copy access list including addresses.
+    let access_list = if let Some(access_list) = state.tx.access_list.clone() {
+        let mut addresses = Vec::with_capacity(access_list.0.len());
+
+        for item in access_list.0.iter() {
+            // Add RW write operations for access list addresses
+            // (will lookup in copy circuit).
+            state.tx_access_list_account_write(exec_step, tx_id, item.address, true, false)?;
+
+            addresses.push((item.address, 0.into()));
+        }
+
+        addresses
+    } else {
+        vec![]
+    };
+
+    let tx_id = NumberOrHash::Number(tx_id);
     let rw_counter_start = state.block_ctx.rwc;
-    state.push_copy(
-        exec_step,
-        CopyEvent {
-            src_addr: 0,
-            src_addr_end: address_size,
-            src_type: CopyDataType::AccessListAddresses,
-            src_id: tx_id.clone(),
-            dst_addr: 0,
-            dst_type: CopyDataType::AccessListAddresses,
-            dst_id: tx_id.clone(),
-            log_id: None,
-            rw_counter_start,
-            // TODO
-            copy_bytes: CopyBytes::new(vec![], None, None),
-        },
-    );
+    let copy_bytes = CopyBytes::new(vec![(0, false, false); access_list.len()], None, None);
+
+    // Add copy event to copy table.
+    let copy_event = CopyEvent {
+        src_type: CopyDataType::AccessListAddresses,
+        src_id: tx_id,
+        src_addr: 1, // index starts from 1.
+        src_addr_end: access_list.len() as u64 + 1,
+        rw_counter_start,
+        copy_bytes,
+        access_list,
+        ..Default::default()
+    };
+
+    state.push_copy(exec_step, copy_event);
+
+    Ok(())
+}
+
+fn add_access_list_storage_key_copy_event(
+    state: &mut CircuitInputStateRef,
+    exec_step: &mut ExecStep,
+) -> Result<(), Error> {
+    let tx_id = state.tx_ctx.id();
+
+    // Build copy access list including addresses and storage keys.
+    let access_list = if let Some(access_list) = state.tx.access_list.clone() {
+        let mut address_storage_keys = vec![];
+
+        for item in access_list.0.iter() {
+            for storage_key in item.storage_keys.iter() {
+                let storage_key = storage_key.to_word();
+
+                // Add RW write operations for access list address storage keys
+                // (will lookup in copy circuit).
+                state.tx_access_list_storage_key_write(
+                    exec_step,
+                    tx_id,
+                    item.address,
+                    storage_key,
+                    true,
+                    false,
+                )?;
+
+                address_storage_keys.push((item.address, storage_key));
+            }
+        }
+
+        address_storage_keys
+    } else {
+        vec![]
+    };
```

Comment:
And this can be similarly refactored.

Reply:
Refactor this code in commit [2501220](https://github.com/scroll-tech/zkevm-circuits/pull/1058/commits/2501220abb87abdf5d7126c4aa3881143f8476b0).